C++: New query 'Failure to use HTTPS URLs'

geoffw0 opened the pull request:
This is a new query for spotting uses of `http` (rather than `https`), which along with another query will provide us a basis to claim partial coverage of [OWASP A08:2021](https://owasp.org/Top10/A08_2021-Software_and_Data_Integrity_Failures/).  The code and `qhelp` is derived from `java/non-https-url` but simplified in several places.

In addition to the tests you can view some results on LGTM at: https://lgtm.com/query/1597019895252712493/

Performance seems fine locally and I have no reason to be concerned about it.

I'm keen to discuss the right precision and severity (and other tags) for this query in addition to any suggestions you have for the query itself.

intrigus-lgtm commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.qhelp:
```diff
@@ -0,0 +1,35 @@
+</li>
+<li>
+OWASP Top 10:
+<a href="https://owasp.org/Top10/A08_2021-Software_and_Data_Integrity_Failures/">A08:2021 � Software and Data Integrity Failures</a>.
```
What's this character in the link text: �

geoffw0 replied:
It's `–`.  I guess I'll replace it with a simple `-`.

geoffw0 replied:
Fixed.  Thanks for spotting.

intrigus-lgtm commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql:
```diff
@@ -0,0 +1,78 @@
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(FunctionCall fc |
+      fc.getTarget().getName() = ["system", "gethostbyname"] and
```
From which libraries are these sinks?

MathiasVP commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql:
```diff
@@ -0,0 +1,78 @@
+    exists(string s | this.getValue() = s |
+      s = "http"
+      or
+      s.matches("http://%") and
+      not s.substring(7, s.length()) instanceof PrivateHostName and
```
Could this not be accomplished in a prettier way using `s.regexpCapture`? Maybe something like:
```ql
exists(string tail | tail = s.regexpCapture("http://(.*)", 1) and not tail instanceof PrivateHostName)
```

geoffw0 replied:
I'm happy with this.  Done.

MathiasVP commented on cpp/ql/test/query-tests/Security/CWE/CWE-319/UseOfHttp/test.cpp:
```diff
@@ -0,0 +1,52 @@
+	openUrl("http://localhost/example"); // GOOD (localhost)
+	openUrl("https://localhost/example"); // GOOD (https, localhost)
+	doNothing("http://example.com"); // GOOD (URL not used)
+	openUrl(urls[0]); // BAD [NOT DETECTED]
```
Is this not detected because `urls` is a global variable? Our support for global variables isn't _great_, and I've been looking for a good motivation to better support it for a while.

geoffw0 replied:
I thought it was because of the array access, but having expanded it into several test cases, the global variable is clearly the issue.

MathiasVP replied:
Fantastic (well, "fantastic" in the sense that it gives me a reason to improve this case)! I'll create an issue for better flow through global variables.

MathiasVP replied:
Created https://github.com/github/codeql-c-team/issues/734 for this.

geoffw0 replied:
Awesome, thanks!

MathiasVP commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql:
```diff
@@ -0,0 +1,78 @@
+      fc.getTarget().getName() = "curl_easy_setopt" and
+      sink.asExpr() = fc.getArgument(2)
+      or
+      fc.getTarget().getName() = "ShellExecute" and
+      sink.asExpr() = fc.getArgument(3)
```
Should this be `fc.getArgument(2)`? Looking at https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecutea, it looks like argument 3 is `lpParameters`?

geoffw0 replied:
I think the most likely offending call would be something along the lines of
```
ShellExecute(hwnd, "open", "browser.exe", "http://example.com", ...);
```
So I expect the URL to be in `lpParameters`.  If it's possible for a URL to be used in other positions I can add those as well.

MathiasVP replied:
Ah, yes. That's a good point. I don't think it's possible to have a URL in other places, so this is probably correct. 👍

MathiasVP commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql:
```diff
@@ -0,0 +1,78 @@
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof HttpStringLiteral }
+
+  override predicate isSink(DataFlow::Node sink) {
```
I'd appreciate it if the QLDoc described what the common thing shared between all the sinks is? Is it "things that do network-related activity when given a URL"?

geoffw0 replied:
It's a bit fuzzier than that, essentially we're looking for evidence that the URL use is not benign.  I've added a comment attempting to explain this.

MathiasVP commented on cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql:
```diff
@@ -0,0 +1,78 @@
+      fc.getTarget().getName() = ["system", "gethostbyname"] and
+      sink.asExpr() = fc.getArgument(0)
+      or
+      fc.getTarget().getName() = ["send", "URLDownloadToFile"] and
```
`send` is a very common function name. Do we not have a model for the particular `send` you're trying to target (or is it a private model that we can't reach from here)?

geoffw0 replied:
It should be `send` in the global namespace, i.e. https://man7.org/linux/man-pages/man2/send.2.html .  For now I've switched `getName` for `hasGlobalOrStdName` but I'll look into upgrading to using a proper model.

geoffw0 replied:
There is a model for `send` and similar functions, but it's private.  I think I should just add this task to the ticket about private models ... except ... I can't find such a ticket.  Do you remember if we actually made it or just committed to doing it elsewhere?  I may have to create a ticket as the problem comes up quite frequently.

MathiasVP replied:
I think we committed to doing it ... somewhere ... but I also can't find any mention of it 🤔 Feel free to create a dedicated issue for it.

geoffw0 replied:
Created a ticket and called out `send` here: https://github.com/github/codeql-c-team/issues/737

MathiasVP reviewed (commented):
Mostly LGTM! None of these are blocking comments.

geoffw0 commented:
I've addressed some of the review comments, and a few more sinks so we get more results: https://lgtm.com/query/2618107714638101401/

More to come...

geoffw0 commented:
All comments addressed.

MathiasVP reviewed (approved):
LGTM! We should probably run a DCA experiment to ensure that we don't get any subtle cache re-evaluation in the extended security suite with this new query.

geoffw0 commented:
DCA run has finished, no new alerts and nothing concerning in the timings as far as I can see.

MathiasVP commented:
Thanks for running DCA Merging 🥳!
